l10n_fi_reconcile/models/account_bank_statement.py

```diff
@@ -0,0 +1,105 @@
+        precision = st_line_currency and st_line_currency.decimal_places or company_currency.decimal_places
+        field = currency and 'amount_residual_currency' or 'amount_residual'
+        liquidity_field = currency and 'amount_currency' or amount > 0 and 'debit' or 'credit'
+        liquidity_amt_clause = currency and '%(amount)s::numeric' or 'abs(%(amount)s::numeric)'
+        params = {'company_id': self.env.user.company_id.id,
+                  'account_payable_receivable': (
+                    self.journal_id.default_credit_account_id.id,
+                    self.journal_id.default_debit_account_id.id),
+                  'amount': float_round(amount, precision_digits=precision),
+                  'partner_id': self.partner_id.id,
+                  'excluded_ids': tuple(excluded_ids),
+                  'ref': self.ref,
+                  }
+
+        if self.ref:
+            # Check for both matching reference and amount
+            add_to_select = ", CASE WHEN LTRIM(aml.payment_reference, '0') = %(ref)s \
+                            THEN 1 ELSE 2 END as temp_field_order "
+            add_to_from = " JOIN account_move m ON m.id = aml.move_id "
+            select_clause, from_clause, where_clause = self._get_common_sql_query(overlook_partner=True, excluded_ids=excluded_ids, split=True)
+            sql_query = select_clause + add_to_select + from_clause + add_to_from + where_clause
+            add_to_where = " AND (LTRIM(aml.payment_reference, '0') = LTRIM(%(ref)s, '0') or m.name = %(ref)s) \
+                            AND (" + field + " = %(amount)s::numeric OR (acc.internal_type = 'liquidity' \
+                            AND " + liquidity_field + " = " + liquidity_amt_clause + ")) \
+                            ORDER BY temp_field_order, date_maturity asc, aml.id asc"
+
+            self.env.cr.execute(sql_query + add_to_where, params)
+            results = self.env.cr.fetchone()
+            if results:
+                return self.env['account.move.line'].browse(results[0])
+
+            # Check for just a matching reference
+            add_to_where = " AND (LTRIM(aml.payment_reference, '0') = LTRIM(%(ref)s, '0') or m.name = %(ref)s) \
+                            ORDER BY temp_field_order, date_maturity asc, aml.id asc"
+            self.env.cr.execute(sql_query + add_to_where, params)
+            results = self.env.cr.fetchone()
+            if results:
+                return self.env['account.move.line'].browse(results[0])
+
+        # Return super if our queries yielded nothing
+        return super(AccountBankStatementLine, self).get_reconciliation_proposition(excluded_ids=excluded_ids)
+
+    def get_statement_line_for_reconciliation_widget(self):
+        res = super(AccountBankStatementLine, self).get_statement_line_for_reconciliation_widget()
+
+        # partner_name key is already in use so use something else
+        if self.partner_name:
+            res['partner_note'] = self.partner_name
```

Comment:
The `partner_name` is a name of a `res.partner` if one exists. Further down we take the `res.partner.name` from an invoice that matches the bank statement line's payment reference.
The self.partner_name is a field imported from the bank statement. It includes some sort of description of the other party of the payment, it might or might not be a name that exists in the system. I wanted to keep the `partner_name` as res.partner name. The bank statement's `partner_name` field was requested to be also visible as a separate column, more as a description field so I added a new key with as descriptive name as I could think of without the name clashing with already existing keys.

The partner_note we can use because it's a completely new key. The returned dict is not used to create a new record so the keys don't have to match any existing fields. The dict is returned to a QWEB view, and I've modified that QWEB to display the value of `partner_note` one a column.
[l10n_fi_reconcile/static/src/xml/account_reconciliation.xml](https://github.com/avoinsystems/l10n-finland/pull/2/files#diff-2b43a6078cfd26eb22d91ce13d065e24)